lib/Sema/ConstraintSystem.cpp

```diff
@@ -3147,25 +3147,60 @@ std::string swift::describeGenericType(ValueDecl *GP, bool includeName) {
 static bool diagnoseConflictingGenericArguments(ConstraintSystem &cs,
                                                 const SolutionDiff &diff,
                                                 ArrayRef<Solution> solutions) {
-  if (!diff.overloads.empty())
+  // Also attempt to detect and diagnose conflicting generic arguments across
+  // multiple solutions and overloads when all fixes are generic mismatches.
+  bool allGenericMismatch =
+      llvm::all_of(solutions, [](const Solution &solution) -> bool {
+        return llvm::all_of(
+            solution.Fixes, [](const ConstraintFix *fix) -> bool {
+              return fix->getKind() == FixKind::GenericArgumentsMismatch;
+            });
+      });
+
+  if (!(diff.overloads.empty() || allGenericMismatch))
     return false;
 
   bool noFixes = llvm::all_of(solutions, [](const Solution &solution) -> bool {
      const auto score = solution.getFixedScore();
      return score.Data[SK_Fix] == 0 && solution.Fixes.empty();
   });
 
-  bool allMismatches =
-      llvm::all_of(solutions, [](const Solution &solution) -> bool {
-        return llvm::all_of(
-            solution.Fixes, [](const ConstraintFix *fix) -> bool {
-              return fix->getKind() == FixKind::AllowArgumentTypeMismatch ||
-                     fix->getKind() == FixKind::AllowFunctionTypeMismatch ||
-                     fix->getKind() == FixKind::AllowTupleTypeMismatch;
-            });
-      });
+  bool allMismatches = false;
+  if (!allGenericMismatch) {
+    allMismatches =
+        llvm::all_of(solutions, [](const Solution &solution) -> bool {
+          // Special casing situations where a type mismatch maybe followed by a
+          // conformance requirement fix on each solution e.g.
+          //    func f<E: Equatable>(_ e1: E, _ e2: E) {}
+          //
+          //    struct A {}
+          //    struct B {}
+          //
+          //    f(A(), B())
+          //
+          // We also have to make sure that we diagnose the generic argument
+          // mismatch in those situations.
```

Comment:
This makes me wonder if a better approach would be no avoid recording conformance failures if generic parameter is already involved in a some other fix? If we do that then we'd end up just looking over the fact that there are requirements on `E` (which don't necessary have to be conformances) and instead pinpoint that arguments for `E` are invalid which makes diagnosing situations like that so much easier...

Reply:
Yeah, I think it would be okay to adjust the score but not record a fix if position whether generic parameter is involved already has one recorded.

Reply:
There is a small problem with the ordering of the fixes are recorded (at least in this case), it seems like the conformance fix is recorded before argument mismatch ... so do you think we should go for not recording the argument mismatch instead if there is a recorded requirement fix for this argument?

Reply:
@xedin I've tried this approach in the last commit, let me know if that is indeed the idea here and also if there is a better way to detect this :) 
But there is a problem that when binding type variable for the opened generic, in this case for `$T0 := A` recording conformance is done before argument fix, and for `B` the other way around, so for the first one there is no way to know if there is going to be another fix involving the generic param ... so although this actually makes this produces a diagnostic for this case is only because one solution is only because it can make one solution better than the other. I still have to check why exactly this happening

Reply:
Right, so this seems to be down to the order in what the constraints are simplified because of the order they are added to the worklist that is the ordering `gatherConstraints` is returning type var constraints. So my understanding right now is for us to ignore only requirement failures we would have to ensure that `ArgumentConversion` constraints are always added to the worklist, therefore, simplified before the `ConformTo` one. So my question is more if we should follow this approach? There are other possible approaches to this? Perhaps go back to the initial one where we handle this on diagnosing ambiguity? Let me know what you think

Reply:
Sorry I wasn't replying, I'm still thinking about how to approach this problem where constraint simplification order in non -deterministic...

Reply:
Ah no worries @xedin :) 
In this time, I've just tried to have more understanding of how this gathering of constraints to activate on a binding step works to try come up with a solution, also experimenting with some things to see how they affect the diagnostics as well. From yesterday to now, I just did the approach where all the fixes are recorded and we try to ignore them at `diagnoseAmbiguity` because it does make sense if we can't rely on skipping fixes. Also, I couldn't come up with an approach to the ordering problem that felt right or at least didn't felt like a hack ... but I may continue looking for a bit more